### PR TITLE
Implement the API changes for getting incremental reconnect updates

### PR DESCRIPTION
**Pull request by tgolen**
- [x] Held on https://github.com/Expensify/Web-Expensify/pull/38303

### Details
When the application reconnects to the network, an API call is made to the server to fetch all data, `reconnectApp()`. This API will now be called with optional `updateIDFrom` and `updateIDTo` parameters. This will tell the server to try and get incremental updates instead of the full amount of data. On the server, if it can't return any incremental data, then it will return a full data set. This saves time and network resources when the app reconnects.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/299548

### Tests
**Note: you must have `define('RELIABLE_UPDATES', true);` defined in web-e in order to test this.**

1. Have two NewDot clients open, signed in with the same user and the application logs open and filtering for `[OnyxUpdates]`
2. On client A, change your preferred language
3. On client B, verify you see logs appear with real `lastUpdateID` and `previousUpdateID` values

- [ ] Verify that no errors appear in the JS console

### Offline tests
None

### QA Steps
Please ping @tgolen and @danieldoglas to QA this. We will have to live-edit staging in order to test it out.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native 
    - [x] Android / Chrome **Note: I have not tested on this platform because the console logs are not available**
    - [x] iOS / native
    - [x] iOS / Safari **Note: I have not tested on this platform because the console logs are not available**
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

**Note: I have not tested on this platform because the console logs are not available**

</details>

<details>
<summary>Mobile Web - Safari</summary>

**Note: I have not tested on this platform because the console logs are not available**

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

image

</details>

<details>
<summary>Android</summary>

image

</details>

### CODE REVIEW

**Comment by tgolen** [comment]
I've removed the HOLD on this. My next steps is to figure out how to test it out and give it a run

**Comment by tgolen** [comment]
This is ready for review now. I have tested it on web/desktop. I need to test it out on native platforms, but ran out of time today, so I'll do that tomorrow.

**Comment by danieldoglas** [comment]
I think this can't be tested by contributors, we need internal testing

**Comment by tgolen** [comment]
OK, I've completed testing on iOS and Android. It helped me to find a proptype warning that I hadn't seen before and I've fixed it, so that's good!

**Review by danieldoglas — DISMISSED** [review]
Code changes LGTM.

@situchan can you test the app with these changes and guarantee it still works?

We need to basically test two separate accounts and guarantee that, to your own accounts (without the beta), you're still able to send messages between them and see things like name updates and etc showing.

**Comment by danieldoglas** [comment]
OK merge conflict is solved. @situchan please review!

**Comment by situchan** [comment]
reviewing now

File: src/libs/Navigation/AppNavigator/AuthScreens.js [hunk 1]
```diff
@@ -127,14 +131,15 @@ class AuthScreens extends React.Component {
         });
 
         // If we are on this screen then we are "logged in", but the user might not have "just logged in". They could be reopening the app
-        // or returning from background. If so, we'll assume they have some app data already and we can call reconnectApp() instead of openApp().
+        // or returning from background. If so, we'll assume they have some app data already and we can call openApp(true) instead of openApp().
         // Note: If a Guide has enabled the memory only key mode then we do want to run OpenApp as their app will not be rehydrated with
         // the correct state on refresh. They are explicitly opting out of storing data they would need (i.e. reports_) to take advantage of
         // the optimizations performed during ReconnectApp.
-        if (this.props.isUsingMemoryOnlyKeys || SessionUtils.didUserLogInDuringSession()) {
+        const shouldGetAllData = this.props.isUsingMemoryOnlyKeys || SessionUtils.didUserLogInDuringSession();
```
Comment (johnmlee101) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:138]: Why this change? Just for clarity?

File: src/libs/actions/App.js [hunk 2]
```diff
@@ -122,63 +122,94 @@ AppState.addEventListener('change', (nextAppState) => {
 });
 
 /**
- * Fetches data needed for app initialization
- * @param {boolean} [isReconnecting]
+ * Gets the policy IDs that are passed to the server in the OpenApp and ReconnectApp API commands. This includes a full list of policy IDs the client knows about as well as when they were last modified.
+ * @returns {Object}
```
Comment (johnmlee101) [inline src/libs/actions/App.js:126]:
```suggestion
 * @returns {Promise}
```

**Review by johnmlee101 — CHANGES_REQUESTED** [review]
minor requests

File: src/libs/Navigation/AppNavigator/AuthScreens.js (hunk 1 above)
Comment (danieldoglas) [reply src/libs/Navigation/AppNavigator/AuthScreens.js:138]: I think so!

File: src/libs/actions/App.js (hunk 2 above)
Comment (danieldoglas) [reply src/libs/actions/App.js:126]: Done

**Comment by situchan** [comment]
I am not getting `lastUpdateID`, `previousUpdateID` from pusher. Is backend deployed to staging or will be?

Screenshot 2023-08-04 at 1 19 33 AM

**Comment by danieldoglas** [comment]
@situchan added that change

**Comment by danieldoglas** [comment]
@situchan @johnmlee101 all yours again!

**Comment by danieldoglas** [comment]
@situchan let me know if you will be able to test this until Monday. If you can't, we'll reassign

**Comment by situchan** [comment]
I will review during weekend

File: src/libs/Navigation/AppNavigator/AuthScreens.js [hunk 3]
```diff
@@ -127,14 +131,15 @@ class AuthScreens extends React.Component {
         });
 
         // If we are on this screen then we are "logged in", but the user might not have "just logged in". They could be reopening the app
-        // or returning from background. If so, we'll assume they have some app data already and we can call reconnectApp() instead of openApp().
+        // or returning from background. If so, we'll assume they have some app data already and we can call openApp(true) instead of openApp().
```
Comment (situchan) [inline src/libs/Navigation/AppNavigator/AuthScreens.js:134]: This change needs to be reverted. After this PR, `openApp` won't have any param.

**Review by situchan — CHANGES_REQUESTED** [review]

File: src/libs/actions/User.js [hunk 4]
```diff
@@ -552,9 +552,22 @@ function subscribeToUserEvents() {
...
             updates = pushJSON;
         } else {
-            // const lastUpdateID = pushJSON.lastUpdateID;
-            // const previousUpdateID = pushJSON.previousUpdateID;
             updates = pushJSON.updates;
+
+            // Not always we'll have the lastUpdateID and previousUpdateID properties in the pusher update
+            // until we finish the migration to reliable updates. So let's check it before actually updating
+            // the properties in Onyx
+            if ('lastUpdateID' in pushJSON && 'previousUpdateID' in pushJSON) {
```
Comment (situchan) [inline src/libs/actions/User.js:560]: NAB
```suggestion
            if (!_.isUndefined(pushJSON.lastUpdateID) && !_.isUndefined(pushJSON.previousUpdateID)) {
```
Or I think this is even better if there's no 0 value for ID:
```suggestion
            if (pushJSON.lastUpdateID && pushJSON.previousUpdateID) {
```

File: src/libs/Navigation/AppNavigator/AuthScreens.js (hunk 3 above)
Comment (danieldoglas) [reply src/libs/Navigation/AppNavigator/AuthScreens.js]: fixed comment

File: src/libs/actions/User.js (hunk 4 above)
Comment (danieldoglas) [reply src/libs/actions/User.js]: Changed usaged.

**Review by situchan — APPROVED** [review]
App works fine as normal.

I was only able to check:
- `[OnyxUpdates] No lastUpdateID and previousUpdateID provided`
- `[OnyxUpdates] App reconnecting with updateIDFrom: 0 and updateIDTo: 0`

These should be tested internally until backend update is available on staging

- `[OnyxUpdates] Received lastUpdateID from pusher`
- `[OnyxUpdates] Received previousUpdateID from pusher`

**Comment by danieldoglas** [comment]
@johnmlee101 or @marcochavezf all yours!

**Review by johnmlee101 — APPROVED** [review]

**Comment by danieldoglas** [comment]
@marcochavezf I'll merge this so we can unblock https://github.com/Expensify/App/pull/23669

**Comment by mvtglobally** [comment]
@tgolen @danieldoglas This is ready to QA

**Comment by danieldoglas** [comment]
@mvtglobally as long the app is working normally (receiving responses from the servers, and loading content), this is tested!

**Comment by danieldoglas** [comment]
These changes are nothing new for now, we just need to guarantee the app is still working.
